pallets/did/src/lib.rs

```diff
@@ -28,91 +30,403 @@ mod tests;
 pub mod benchmarking;
 
 pub mod default_weights;
-pub mod migration;
 pub use default_weights::WeightInfo;
 
 use codec::{Decode, Encode};
-use frame_support::{decl_event, decl_module, decl_storage, dispatch::DispatchResult, Parameter, StorageMap};
+
+use codec::EncodeLike;
+use frame_support::{ensure, storage::types::StorageMap};
 use frame_system::{self, ensure_signed};
-use sp_runtime::{codec::Codec, traits::Member};
-use sp_std::prelude::*;
+use sp_core::{ed25519, sr25519};
+use sp_runtime::traits::Verify;
+use sp_std::{collections::btree_set::BTreeSet, convert::TryFrom, fmt::Debug, prelude::Clone, vec::Vec};
+
+pub use pallet::*;
+
+/// Reference to a payload of data of variable size.
+pub type PayloadReference<'a> = &'a [u8];
+
+/// Type of a signature (variable size as different signature schemes are
+/// supported).
+pub type SignatureEncoding = Vec<u8>;
 
-/// The DID trait
-pub trait Config: frame_system::Config {
-	/// DID specific event type
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+/// Reference to a signature of variable size.
+pub type SignatureReference<'a> = &'a [u8];
 
-	/// Weight information for extrinsics in this pallet.
-	type WeightInfo: WeightInfo;
+/// Type for an encoded URL.
+pub type UrlEncoding = Vec<u8>;
```

Comment:
That feels odd. Having two types for that. 🤔 
 
I would replace `pub type PayloadReference<'a> = &'a [u8];` with `pub type PayloadSlice = [u8];` since it's a slice. You would need to use it then as `&PayloadSlice` to keep it as a reference.

Reply:
It's hard to avoid having two types here. 

But i would prefer an enum of ed25519 or sr25519 signatures instead of an `[u8]`. The signatures can than also not have the wrong size.